src/ZkFold/Base/Algebra/Polynomials/Multivariate.hs

```diff
@@ -28,35 +40,78 @@ import           ZkFold.Base.Algebra.Polynomials.Multivariate.Monomial
 import           ZkFold.Base.Algebra.Polynomials.Multivariate.Polynomial
 import           ZkFold.Base.Algebra.Polynomials.Multivariate.Set
 import           ZkFold.Base.Algebra.Polynomials.Multivariate.Substitution
+import           ZkFold.Prelude                                            (elemIndex)
 
 -- | Most general type for a multivariate monomial
-type SomeMonomial = M Natural Natural (Map Natural Natural)
+type Monomial' = M Natural Natural (Map Natural Natural)
 
 -- | Most general type for a multivariate polynomial
-type SomePolynomial c = P c Natural Natural (Map Natural Natural) [(c, M Natural Natural (Map Natural Natural))]
+type Polynomial' c = P c Natural Natural (Map Natural Natural) [(c, Monomial')]
 
 -- | Monomial constructor
 monomial :: Monomial i j => Map i j -> M i j (Map i j)
 monomial = M . fromJust . toMonomial
 
 -- | Polynomial constructor
-polynomial :: Polynomial c i j => [(c, M i j (Map i j))] -> P c i j (Map i j) [(c, M i j (Map i j))]
-polynomial = sum . map (\m -> P [m]) . fromJust . toPolynomial
+polynomial ::
+    Polynomial c i j =>
+    [(c, M i j (Map i j))] ->
+    P c i j (Map i j) [(c, M i j (Map i j))]
+polynomial = sum . map (\m -> P [m]) . toPolynomial
 
 -- | @'var' i@ is a polynomial \(p(x) = x_i\)
 var :: Polynomial c i j => i -> P c i j (Map i j) [(c, M i j (Map i j))]
 var x = polynomial [(one, monomial (singleton x one))]
 
-evalMonomial :: forall i j m b . (FromMonomial i j m, MultiplicativeMonoid b, Exponent b j) => (i -> b) -> M i j m -> b
-evalMonomial f (M m) = product (map (\(i, j) -> f i ^ j) (toList $ fromMonomial @i @j m))
+evalMonomial :: forall i j m b .
+    FromMonomial i j m =>
+    MultiplicativeMonoid b =>
+    Exponent b j =>
+    (i -> b) -> M i j m -> b
+evalMonomial f (M m) = product
+    $ toList (fromMonomial @i @j m)
+        <&> (\(i, j) -> f i ^ j)
 
-evalPolynomial :: forall c i j m p b . (FromMonomial i j m, FromPolynomial c i j m p, Algebra c b, Exponent b j)
-    => (i -> b) -> P c i j m p -> b
-evalPolynomial f (P p) = sum $ map (\(c, m) -> scale c (evalMonomial f m)) (fromPolynomial @c @i @j @m @p p)
+evalPolynomial :: forall p c i j m b .
+    FromMonomial i j m =>
+    FromPolynomial c i j m p =>
+    Algebra c b =>
+    Exponent b j =>
+    (i -> b) -> P c i j m p -> b
+evalPolynomial f (P p) = sum
+    $ fromPolynomial @c @i @j @m @p p
+        <&> (\(c, m) -> scale c $ evalMonomial f m)
 
-variables :: forall c i j m p . (FromMonomial i j m, FromPolynomial c i j m p) => P c i j m p -> [i]
-variables (P p) = nubOrd $ concatMap (\(_, M m) -> keys (fromMonomial @i @j @m m)) $ fromPolynomial @c @i @j @m @p p
+variables :: forall c i j m p .
+    FromMonomial i j m =>
+    FromPolynomial c i j m p =>
+    P c i j m p -> [i]
+variables (P p) = nubOrd
```

Comment:
I do similar thing using `Sources` newtype from `MonadBlueprint` module, maybe we can improve code reusage by using it (and moving it from `MonadBlueprint` somewhere else)?

Reply:
@TurtlePU Probably, but they have different types. in case of `sources` it's CPS encoding for witnesses.

Reply:
I mean, you can receive the same result by running `evalPolynomial` and picking `b` to be `Sources c i`

Reply:
Of course, no prob!